lib/taxonomy/govuk_taxonomy.rb

```diff
@@ -21,10 +21,39 @@ def all_taxons
       @_all_taxons ||= children.flat_map(&:tree) + draft_child_taxons.flat_map(&:tree)
     end
 
+    def matching_against_published_taxons(taxons)
+      @_published_taxons ||= matching_against_taxonomy_branches(taxons, children)
+    end
+
+    def matching_against_visible_draft_taxons(taxons)
+      @_draft_visible_taxons ||= matching_against_taxonomy_branches(taxons, draft_child_taxons)
+    end
+
   private
 
     def build_tree(taxon_hash)
       @tree_builder_class.new(taxon_hash).root_taxon
     end
+
+    def matching_against_taxonomy_branches(taxons, taxononomy_branches)
+      matching_taxons = []
+      taxononomy_branches.each do |branch|
+        matched_taxons = filter_against_taxonomy_branch(taxons, branch, [])
+        matching_taxons.concat(matched_taxons) if matched_taxons.any?
+      end
+      matching_taxons
+    end
+
+    def filter_against_taxonomy_branch(selected_taxons_content_ids, taxon, matched_taxons)
+      if selected_taxons_content_ids.include?(taxon.content_id)
+        matched_taxons << taxon.content_id
+      end
+
+      taxon.children.each do |child_taxon_branch|
+        filter_against_taxonomy_branch(selected_taxons_content_ids, child_taxon_branch, matched_taxons)
+      end
+
+      matched_taxons
```

Comment:
This mutates the `matched_taxons` array if I'm not mistaken, which is why I guess you can get away with not using the return value when doing the recursive call.

Mutation isn't great in my opinion, as it often makes code unobviously complex, but this is private, so it's not too bad here. Especially using `flat_map`, I don't think this would be too difficult to write avoiding mutation, but I'll leave the decision up to you.